File: Classification/THRESHOLDerrorMinimizer/R/THRESHOLDerrorMinimizer.R

`THRESHOLDerrorMinimizer` <-
function(data, labs, weights, direction = "b")
{
	#cat("Original labs:", labs, "\n")
	#cat("Original weights:", weights, "\n")
    
    geq <- 1
    leq <- 1
    if(direction == "l")
    {
        geq <- 0
    }
    
    if(direction == "g")
    {
        leq <- 0
    }

	numFea <- nrow(data)  #Number of Samples 

	posIndex <- which(labs==1)
	negIndex <- which(labs==0)

	num_pos	<- length(posIndex)
	num_neg <- length(negIndex)

	#cat("Sorting data\n")
	data <- data[, c(posIndex, negIndex), drop = FALSE]

	#cat("Sorting weights\n")
	sorted.weights <- weights[c(posIndex, negIndex)]

	#print(sorted.weights)

	sorted.feature <- double(ncol(data))
	
	#cat("generating normVecs\n")
	thresholds <- double(ncol(data)+1)
	
	#print(normVecs[1:10,])

	allResults = double(4 * numFea)
	#cat("Starting C code\n")

	result <- .C("THRESHOLDerrorMinimizer",
	   as.double(t(data)),					#Dataset (sorted) 
	   as.double(sorted.weights),			#Weigths (sorted)
	   as.double(sorted.feature),			#Feature (sorted)
	   as.double(thresholds),				#List of thresholds
	   as.integer(numFea),					#Number of Features
	   as.integer(num_pos),					#Index of last positive example
	   as.integer(ncol(data)),				#Index of last negative example
	   as.integer(length(thresholds)),		#Index of last thresholds
       as.integer(geq),                     #Allow >= threshold (Yes = 1, No = Else)
       as.integer(leq),                     #Allow <= threshold (Yes = 1, No = Else)
	   allResults = allResults)$allResults  #The list of all results
	
	#cat("Stopping C code\n")
	
	result <- t(matrix(result,nrow = 4))
	colnames(result) <- c("Fea", "Dir", "Thr","Score")

	result[,1] <- result[,1]+1
	

	return(list(
			base.learners = result,
			weights	      = weights,
			type	      = "minimize"
	))	
}


`baselearner.threshold` <- function(data, labs,type = "error", weights = NULL, direction = "b")
{
	if(type == "error")
	{
		numSam <- length(labs)
		return(THRESHOLDerrorMinimizer(data, labs, rep(1/numSam,numSam), direction))
	} 
	if(type == "classwise")
	{
		indexPos <- labs==1
		indexNeg <- labs==0

		weights <- rep(0, length(labs))
		weights[indexPos] <- 1/(sum(indexPos)*2)
		weights[indexNeg] <- 1/(sum(indexNeg)*2)

		return(THRESHOLDerrorMinimizer(data, labs, weights, direction))
	}
	if(type == "weighted")
	{
		if(is.null(weights))
		    stop("Error in function baselearner.threshold: Argument weights is empty.")

		weights =weights/sum(weights)
		return(THRESHOLDerrorMinimizer(data, labs, weights, direction))
	}

        stop("Error in function baselearner.threshold: Unknown type selected.")
}



`prediction.threshold` <- function(models,newdata)
{
	numFea <- nrow(newdata)
	numSam <- ncol(newdata)

	features   <- models[,"Fea"]-1
	directions <- models[,"Dir"]
	thresholds <- models[,"Thr"]

	numMod <- nrow(models)

	if(max(features)+1 > numFea)
		stop("Trained classification models do not fit to newdata: Unknown feature dimension selected.")

	allResults <- rep(0, numMod*numSam )

	result <- .C("THRESHOLDprediction",
	   as.double(t(newdata)),				#Dataset (sorted) 
	   as.integer(features),				#Model parameter: Feature
	   as.integer(directions),				#Model parameter: Direction
	   as.double(thresholds),				#Model parameter: Threshold
	   as.integer(numFea),					#Number of Features (newdata)
	   as.integer(numSam),					#Number of Samples (newdata)
	   as.integer(numMod),					#Number of Modles (models)
	   allResults = as.double(allResults))$allResults	#The list of all results    

        result <- t(matrix(result, nrow = numSam, ncol = numMod ))

	return(result)
}

File: Classification/THRESHOLDerrorMinimizer/src/THRESHOLDerrorMinimizer.c
#include "THRESHOLDerrorMinimizer.h"


void THRESHOLDerrorMinimizer(	double* data,
				double* weights,  
				double*	sortedFeature,
				double*	thresholds,  
				int* 	numFea,
				int* 	last_pos, 
				int* 	last_neg, 
				int*	numThresh,
                int*    testGEQ,
                int*    testLEQ,
                double* output)
{

//counter

int i;	//Iterator Thresholds
int j;	//Iterator Samples

int k;	//Iterator of feature



double* actWeight;	// Current weight

double* feature;	// Pointer to feature dimension

double* actSortedFea;

double* actFea;		// Pointer to current sample feature dimension

double* thresh;
double* actThreshold;

double* actOutput;	// Pointer to current position in output

double	actScore;	// Score of current classifier

double  bestScore;	 // Best score for feature combination
double  bestDir;	 // Begin of best interval (for feature combination)
double  bestThr;         // End of best interval (for feature combination)


//int 	verbose;


//verbose=0;

actOutput= output;

thresh = thresholds;

actSortedFea = sortedFeature;

	for (k =0; k < *numFea; k++)
	{
		feature      = &data[*last_neg*k];
		actSortedFea = sortedFeature;
		
		for(i=0; i < *last_neg; i++)
		{
			*actSortedFea++ = *feature++;
		}

		feature      = &data[*last_neg*k];
		actSortedFea = sortedFeature;

		R_rsort (actSortedFea,*last_neg);


		//printf("************************************\n");
		for(i=0; i < (*last_neg-1); i++)
		{
			thresh[i] = (actSortedFea[i] + actSortedFea[i+1])/2.0;
		//	printf("%f \n",thresh[i]);
		}
		
		thresh[(*last_neg-1)] = actSortedFea[0]-0.5;
		thresh[*last_neg]     = actSortedFea[(*last_neg-1)]+0.5;

		

		actThreshold = thresh;

		bestScore = 99999999999999999.9;
		bestDir   = 1.0;
		bestThr   = -1.0;    

      
        for(i=0; i < *numThresh; i++)
        {
            if(*testGEQ == 1)
            {
                //****************************************************
                //**
                //** Analyse positive direction

                actFea   = feature;
                actWeight = weights;

                actScore = 0;

                // Classify positive examples
                for (j =0; j < *last_pos; j++)
                {
                    //Misclassified Sample of class 1
                    if( *actFea++ <= *actThreshold)
                    {
                        actScore = actScore + (*actWeight);
                    }
		
                    actWeight++;
                }

                // Classify negative examples
                for (j =*last_pos; j < *last_neg; j++)
                {
                    if( *actFea++ > *actThreshold)
                    {
                        actScore = actScore + (*actWeight);
                    }

                    actWeight++;
                }

                // Updating currentComb and bestScoreComb
                if(actScore < bestScore)
                {
                    bestScore = actScore;
                    bestDir   = 1;
                    bestThr   = *actThreshold;
                    //printf("%f \n",bestThr);
                }
            }

            if(*testLEQ == 1)
            {
                //****************************************************
                //**
                //** Analyse negative direction

                actFea    = feature;
                actWeight = weights;

                actScore = 0;

                // Classify positive examples
                for (j =0; j < *last_pos; j++)
                {
                    //Misclassified Sample of class 1
                    if( *actFea++ >= *actThreshold)
                    {
                        actScore = actScore + (*actWeight);
                    }
		
                    actWeight++;
                }

                // Classify negative examples
                for (j =*last_pos; j < *last_neg; j++)
                {
                    if( *actFea++ < *actThreshold)
                    {
                        actScore = actScore + (*actWeight);
                    }

                    actWeight++;
                }

                // Updating currentComb and bestScoreComb
                if(actScore < bestScore)
                {
                    bestScore = actScore;
                    bestDir   = -1.0;
                    bestThr   = *actThreshold;
                    //printf("%f \n",bestThr);
                }
            }

			actThreshold++;
        }

		*actOutput++ = (double) k;//feature
		*actOutput++ = bestDir;//direction
		*actOutput++ = bestThr;//threshold
		*actOutput++ = bestScore;//score
	}//END for (k =0; k < (*numFea)-1; k++)

}

void THRESHOLDprediction(	double* data,
				int*    features, 
				int*    directions, 
				double*	thresholds, 
				int* 	numFea, 
				int* 	numSam, 
				int* 	numMod, 
				double* output)
{
	int i;
	int j;	

	double* actData;
	int* 	actFea;
	int*	actDir;
	double* actThr;
	double* actOut;

	actFea = features;
	actDir = directions;
	actThr = thresholds;
	actOut = output;

	for(i=0; i < *numMod; i++)
	{
		actData = &data[(*actFea++)*(*numSam)];

		if(*actDir++ == 1)
		{
			for(j =0; j < *numSam;j++)
			{
				if(*actData++ > *actThr)
				{
					*actOut = 1.0;
				}
				actOut++;
			}
		}else{
			for(j =0; j < *numSam;j++)
			{
				if(*actData++ < *actThr)
				{
					*actOut = 1.0;
				}
				actOut++;
			}
		}

		actThr++;
	}
}
